pkg/generator/rps_test.go

```diff
@@ -0,0 +1,376 @@
+			},
+			expectedCount: []int{8, 7},
+		},
+	}
+
+	epsilon := 0.01
+
+	for _, test := range tests {
+		t.Run("warm_start"+test.testName, func(t *testing.T) {
+			matrix, minuteCount := GenerateWarmStartFunction(test.experimentDuration, test.rpsTarget)
+
+			if len(matrix) != len(test.expectedIAT) {
+				t.Errorf("Unexpected IAT array size - got: %d, expected: %d", len(matrix), len(test.expectedIAT))
+			}
+			if len(minuteCount) != len(test.expectedCount) {
+				t.Errorf("Unexpected count array size - got: %d, expected: %d", len(minuteCount), len(test.expectedCount))
+			}
+
+			sum := 0.0
+			count := 0
+			currentMinute := 0
+
+			for i := 0; i < len(matrix); i++ {
+				if math.Abs(matrix[i]-test.expectedIAT[i]) > epsilon {
+					t.Error("Unexpected IAT value.")
+				}
+
+				sum += matrix[i]
+				count++
+
+				if int(sum/60_000_000) != currentMinute {
+					if count != test.expectedCount[currentMinute] {
+						t.Error("Unexpected count array value.")
+					}
+
+					currentMinute = int(sum / 60_000_000)
+					count = 0
+				}
+			}
+		})
+	}
+}
+
+func TestColdStartMatrix(t *testing.T) {
+	tests := []struct {
+		testName           string
+		experimentDuration int
+		rpsTarget          float64
+		cooldownSeconds    int
+		expectedIAT        []common.IATArray
+		expectedCount      [][]int
+	}{
+		{
+			testName:           "2min_1rps",
+			experimentDuration: 2,
+			rpsTarget:          1,
+			cooldownSeconds:    10,
+			expectedIAT: []common.IATArray{
+				{10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000},
+				{11_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000},
+				{12_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000},
+				{13_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000},
+				{14_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000},
+				{15_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000},
+				{16_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000},
+				{17_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000},
+				{18_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000},
+				{19_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000, 10_000_000},
```

Comment:
Doesn't see right. Sum of all of them is 129s, which is bigger than expected 2min.

Reply:
RPS mode works slightly differently. Concretely, this is a 100% cold start experiment with RPS=1. Because Knative's minimum autoscaling stable window is 6s, the minimum keep-alive for a function is 6s. This means that we need multiple functions to achieve RPS=1, each scaling up/and down with a 1-second delay. In RPS mode, the number of functions is determined by the `RpsCooldownSeconds` parameter. This produces the IAT array as above. Due to the math complexity, the cold start experiment sleeps for the first ten seconds. In the results, the user should discard the first and the last `RpsCooldownSeconds` of the results, since the RPS at those points is lower than the requested one.

I will add the paragraph above to the documentation and to the comments in the code.